src/Snak/SnakObject.php

```diff
@@ -116,4 +125,46 @@ public function unserialize( $serialized ) {
 		$this->propertyId = PropertyId::newFromNumber( unserialize( $serialized ) );
 	}
 
+	/**
+	 * @param string $name
+	 *
+	 * @return boolean
+	 */
+	public function hasFacet( $name ) {
+		return $this->facetManager && $this->facetManager->hasFacet( $name );
+	}
+
+	/**
+	 * @return string[]
+	 */
+	public function listFacets() {
+		return $this->facetManager ? $this->facetManager->listFacets() : array();
+	}
+
+	/**
+	 * @param string $name
+	 * @param string|null $type The desired type
+	 *
+	 * @return object
+	 */
+	public function getFacet( $name, $type = null ) {
+		if ( !$this->facetManager ) {
+			throw new NoSuchFacetException( $name );
+		}
+
+		return $this->facetManager->getFacet( $name, $type );
+	}
+
+	/**
+	 * @param string $name
+	 * @param object $facetObject
+	 */
+	public function addFacet( $name, $facetObject ) {
+		if ( !$this->facetManager ) {
+			$this->facetManager = new FacetManager();
+		}
+
+		$this->facetManager->addFacet( $name, $facetObject );
```

Comment:
We need to switch to PHP 5.4 first before being able to use traits. And I remember hearing opinions that we should probably never use traits, because they do not solve the issues abstract base classes introduce.